jcl/src/java.base/share/classes/java/lang/String.java

```diff
@@ -3199,49 +3204,41 @@ public boolean regionMatches(int thisStart, String string, int start, int length
 			return false;
 		}
 
+		return regionMatchesInternal(s1, s2, s1.value, s2.value, thisStart, start, length);
+	}
+
+	/*[IF Sidecar19-SE]*/
+	private static boolean regionMatchesInternal(String s1, String s2, byte[] s1Value, byte[] s2Value, int s1Start, int s2Start, int length)
+	/*[ELSE]*/
+	private static boolean regionMatchesInternal(String s1, String s2, char[] s1Value, char[] s2Value, int s1Start, int s2Start, int length)
+	/*[ENDIF]*/
+	{
 		if (length <= 0) {
 			return true;
 		}
-
-		int o1 = thisStart;
-		int o2 = start;
-
 		// Index of the last char to compare
 		int end = length - 1;
-
-		/*[IF Sidecar19-SE]*/
-		byte[] s1Value = s1.value;
-		byte[] s2Value = s2.value;
-		/*[ELSE]*/
-		char[] s1Value = s1.value;
-		char[] s2Value = s2.value;
-		/*[ENDIF]*/
-
-		s1Value.getClass(); // Implicit null check
-		s2Value.getClass(); // Implicit null check
```

Comment:
Will performance regress if we remove these? Wouldn't we get null checks on the access of `s1Value` in every iteration of the loop?

Reply:
Uups, meant to post a review comment but I accepted accidentally. Treat this as a "Review Comment".

Reply:
`string.value` can never be null, therefore these null checks are unnecessary. The optimizer should not generate any null checks for any usage of `string.value`; and hence removing it does not negatively impact its performance.

Reply:
Is the optimizer able to track that `s1Value` local variable (argument) is really an array that came from a String object?

Reply:
I believe the optimizer is capable to track `s1Value` provided the callees are inlined.